src/contracts/GPv2Settlement.sol

```diff
@@ -28,8 +23,16 @@ contract GPv2Settlement {
             chainId := chainid()
         }
 
-        replayProtection = keccak256(
-            abi.encode(DOMAIN_SEPARATOR, chainId, address(this))
+        domainSeparator = keccak256(
+            abi.encode(
+                keccak256(
+                    "EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)"
+                ),
+                keccak256("Gnosis Protocol"),
+                keccak256("v2"),
+                chainId,
+                address(this)
```

Comment:
Isn't omitting the salt different from not setting its value (leave it at 0x00..00). 

As for the structure, I was expecting we would declare

```solidity
struct EIP712Domain {
  string: name,
  string: version
  ...
}
```

and then in the constructor fill an instance of it and abi encode it. 

I'm also surprised that we hash the name and version string instead of setting it as its full length string (the unit tests also suggest otherwise). That being said I'm not sure how important it is to fully follow this standard in practice (since I'm really nitpicky here).

PS: Here is the [example implementation](https://github.com/ethereum/EIPs/blob/master/assets/eip-712/Example.sol) from the EIP where they also omit the salt, so I guess that's fine.

Reply:
> Isn't omitting the salt different from not setting its value

The spec specifies that all values are optional. What changes is the type hash. So, if you only include the name and chain ID, then the type hash would be `keccak("EIP712Domain(string name,uint256 chainId)")`, If you omit all domain fields (which would still be valid!) then it would be `keccak("EIP712Domain()")`.

The fields in the struct are all `abi.encoded` in the order they appear in the type hash (a well-defined order in the spec). Notably, the fields are encoded such that they all fit in a word. In that sense, dynamic sized types like `string` and `bytes` aren't directly `abi.encode`-ed directly, but instead their keccak hash is used intead. From the spec in the [`encodeData` section](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-712.md#definition-of-encodedata):

> The dynamic values bytes and string are encoded as a keccak256 hash of their contents.

This is why, when encoding the field data for name and version, we hash the string content instead of just the string.

> the unit tests also suggest otherwise

The `ethers` library [does this internally](https://github.com/ethers-io/ethers.js/blob/e0b1d0e1e0280af860aaa593744a1a177f828397/packages/hash/src.ts/typed-data.ts#L133-L138).

> That being said I'm not sure how important it is to fully follow this standard in practice (since I'm really nitpicky here).

Super important as we want to add support for `eth_signTypedData` in #143. These are all great questions that took me a long time to figure out :stuck_out_tongue:

Reply:
Note that we have a test to ensure that the computed EIP712 domain separator is correctly computed, or at least that it matches the Ethers.js implementation:

https://github.com/gnosis/oba-contracts/blob/fc385496379537f84c46cef0e252ad873abcecd1/test/GPv2Settlement.test.ts#L23-L27